### PR TITLE
Support buzzer on secondary board

### PR DESCRIPTION
**Pull request by qwandor**
This won't work properly until I fix a bug in `gd32f1x0-hal` (in https://github.com/qwandor/gd32f1x0-hal/pull/5), but the code on this side is good at least.

### CODE REVIEW

File: cross/hoverkite-firmware/src/protocol.rs [hunk 1]
```diff
@@ -193,6 +193,14 @@ pub fn handle_command(
                 hoverboard.leds.green.set_low().unwrap()
             }
         }
+        Command::SetBuzzerFrequency(frequency) => {
+            log!(hoverboard.response_tx(), "Buzzer {} Hz", frequency);
+            hoverboard.buzzer.set_frequency(if frequency == 0 {
```
Comment (alsuren) [inline cross/hoverkite-firmware/src/protocol.rs:198]: I'd prefer for this command to have a duration argument and set a timer to interrupt to disable to buzzer after that time expires. This way there's less chance of accidentally setting off a forever-alarm.

`Beep { frequency, duration }`?

File: cross/hoverkite-firmware/src/hoverboard.rs [hunk 2]
```diff
@@ -150,12 +153,50 @@ fn DMA_CHANNEL0() {
...
 
+/// The buzzer on the secondary board. This should not be used on the primary board.
+pub struct Buzzer {
+    pwm: pwm::Pwm<
+        TIMER1,
+        // This type is a bit bogus, PB10 is the only one that matters.
+        (
+            Option<PA0<Alternate<AF2>>>,
+            Option<PA1<Alternate<AF2>>>,
+            Option<PB10<Alternate<AF2>>>,
```
Comment (alsuren) [inline cross/hoverkite-firmware/src/hoverboard.rs:164]: could/should the pwm module export a second type for single-pin pwm that's simpler?

File: cross/hoverkite-firmware/src/protocol.rs (hunk 1 above)
Comment (qwandor) [reply cross/hoverkite-firmware/src/protocol.rs:202]: Added `Command::AddBuzzerNote`.

File: messages/src/command.rs [hunk 3]
```diff
@@ -3,17 +3,38 @@ use crate::util::{ascii_to_bool, bool_to_ascii};
...
+use core::fmt::{self, Display, Formatter};
 use core::mem::size_of;
 use core::ops::RangeInclusive;
 use nb::Error::{Other, WouldBlock};
 
+/// A note to play on the buzzer.
+#[derive(Clone, Copy, Debug, Default, Eq, PartialEq)]
+pub struct Note {
+    /// The frequency in Hertz, or `None` for silence.
+    pub frequency: Option<u32>,
```
Comment (alsuren) [inline messages/src/command.rs:15]: `Option<NonZero<u32>>` uses fewer bytes and encodes what you mean better

**Review by alsuren — APPROVED** [review]

File: cross/hoverkite-firmware/src/main.rs [hunk 4]
```diff
@@ -98,6 +104,9 @@ fn main() -> ! {
     let mut target_position: Option<i64> = None;
     let mut speed_limits = -200..=200;
     let mut spring_constant = 10;
+    let mut note_queue = CircularBuffer::default();
```
[the remark this answers is not in the document]
Comment (alsuren) [reply cross/hoverkite-firmware/src/main.rs:107]: ah. It's in the function you're calling, below. How difficult is it to make that function take a const usize SIZE generic parameter

[the remark this answers is not in the document]
Comment (qwandor) [reply cross/hoverkite-firmware/src/main.rs]: Done.

File: messages/src/command.rs (hunk 3 above)
Comment (qwandor) [reply messages/src/command.rs]: Aha, I was wondering about something like that. Done.

File: cross/hoverkite-firmware/src/circular_buffer.rs [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+impl<T: Copy + Default, const SIZE: usize> CircularBuffer<T, SIZE> {
+    /// Try to add the given element to the buffer. Returns true on success, or false if the buffer
+    /// was already full.
+    pub fn add(&mut self, element: T) -> bool {
+        if self.length == self.buffer.len() {
+            return false;
+        }
+        self.buffer[(self.start + self.length) % self.buffer.len()] = element;
```
[the remark this answers is not in the document]
Comment (qwandor) [reply cross/hoverkite-firmware/src/circular_buffer.rs]: Done.

File: cross/hoverkite-firmware/src/circular_buffer.rs [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            start: 0,
+            length: 0,
+        }
+    }
+}
+
+impl<T: Copy + Default, const SIZE: usize> CircularBuffer<T, SIZE> {
+    /// Try to add the given element to the buffer. Returns true on success, or false if the buffer
+    /// was already full.
+    pub fn add(&mut self, element: T) -> bool {
```
[the remark this answers is not in the document]
Comment (qwandor) [reply cross/hoverkite-firmware/src/circular_buffer.rs:32]: Done.

File: cross/hoverkite-firmware/src/circular_buffer.rs [hunk 7]
```diff
@@ -0,0 +1,84 @@
+/// A circular buffer.
+pub struct CircularBuffer<T: Default, const SIZE: usize> {
```
[the remark this answers is not in the document]
Comment (qwandor) [reply cross/hoverkite-firmware/src/circular_buffer.rs]: Done.

**Pull request merged**